homeassistant/components/metoffice/__init__.py

```diff
@@ -1 +1,106 @@
-"""The metoffice component."""
+"""The Met Office integration."""
+
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE, CONF_NAME
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    METOFFICE_COORDINATOR,
+    METOFFICE_DATA,
+    METOFFICE_NAME,
+)
+from .data import MetOfficeData
+
+_LOGGER = logging.getLogger(__name__)
+
+_METOFFICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Inclusive(
+            CONF_LATITUDE, "coordinates", "Latitude and longitude must exist together"
+        ): cv.latitude,
+        vol.Inclusive(
+            CONF_LONGITUDE, "coordinates", "Latitude and longitude must exist together"
+        ): cv.longitude,
+        vol.Required(CONF_NAME): cv.string,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [_METOFFICE_SCHEMA])}, extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORMS = ["weather", "sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the National Weather Service (NWS) component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up a Met Office entry."""
+
+    latitude = entry.data[CONF_LATITUDE]
+    longitude = entry.data[CONF_LONGITUDE]
+    api_key = entry.data[CONF_API_KEY]
+    name = entry.data[CONF_NAME]
+
+    metoffice_data = MetOfficeData(hass, api_key, latitude, longitude)
+    await metoffice_data.async_update_site()
```

Comment:
What happens if this fails here but not in the config_flow?  I think `ConfigEntryNotReady` can be raised.  Ideally, we could set up the data inside the config_flow and pass it here, but I couldn't figure that out.

Reply:
That makes sense - I'll give it a go and see what I can do with that.

Reply:
As you say, you can't just pass back the data object (as it can't be easily converted to/from JSON) so I now check the newly created data object for a site name and weather data and raise the exception if either fails.

Reply:
I was hoping that a core dev would see the redundancy and extra problems here and suggest a good method to handle it in future PRs for both of us

Reply:
Yeah, I looked at whether to try and add methods to serialize the ...Data object thru JSON back into the config_flow handler, but it seemed kinda hokey when all that would do would be to save my code the trouble of actually creating the instance itself (as essentially it would be creating a new instance from the same api_key, latitude and longitude)... overkill? Felt like it.